openfe/setup/network.py

```diff
@@ -53,6 +56,90 @@ def nodes(self) -> FrozenSet[Molecule]:
         """A read-only view of the nodes of the Network"""
         return self._nodes
 
+    def __eq__(self, other):
+        return self.nodes == other.nodes and self.edges == other.edges
+
+    def _serializable_graph(self) -> nx.Graph:
+        """
+        Create NetworkX graph with serializable attribute representations.
+
+        This enables us to use easily use different serialization
+        approaches.
+        """
+        # sorting ensures that we always preserve order in files, so two
+        # identical networks will show no changes if you diff their
+        # serialized versions
+        sorted_nodes = sorted(self.nodes, key=lambda m: (m.smiles, m.name))
+        mol_to_label = {mol: f"mol{num}"
+                        for num, mol in enumerate(sorted_nodes)}
+
+        edge_data = sorted([
+            (
+                mol_to_label[edge.mol1],
+                mol_to_label[edge.mol2],
+                json.dumps(list(edge.mol1_to_mol2.items()))
+            )
+            for edge in self.edges
+        ])
+
+        # from here, we just build the graph
+        serializable_graph = nx.MultiDiGraph(ofe_version=openfe.__version__)
+        for mol, label in mol_to_label.items():
+            serializable_graph.add_node(label, sdf=mol.to_sdf())
+
+        for mol1, mol2, mapping in edge_data:
+            serializable_graph.add_edge(mol1, mol2, mapping=mapping)
+
+        return serializable_graph
+
+    @classmethod
+    def _from_serializable_graph(cls, graph: nx.Graph):
+        """Create network from NetworkX graph with serializable attributes.
+
+        This is the inverse of ``_serializable_graph``.
+        """
+        label_to_mol = {node: Molecule.from_sdf_string(sdf)
+                        for node, sdf in graph.nodes(data='sdf')}
+
+        edges = [
+            AtomMapping(mol1=label_to_mol[node1],
+                        mol2=label_to_mol[node2],
+                        mol1_to_mol2=dict(json.loads(mapping)))
+            for node1, node2, mapping in graph.edges(data='mapping')
+        ]
+
+        return cls(edges=edges, nodes=label_to_mol.values())
```

Comment:
I'd probably just inline these methods until we have a use for the different entrypoint

Reply:
I see the YAGNI argument, but to be honest, I'm not 100% clear on the relative importance of various graph serialization formats. I seemed to me like GraphML and GEXF were most important, and I went with GraphML as the more software-independent solution. But I'd like to leave open the possibility of rapidly changing that if needed, because I'm definitely not an expert in this space.

Separated like this, we (or even a brave user) can immediately use any of the [other formats](https://networkx.org/documentation/stable/reference/readwrite/index.html) that NetworkX can read/write -- at least, any of the formats that preserve node/edge attributes. If you feel strongly about inlining, I can do that, but this is why I didn't inline originally.